warehouse/accounts/security_policy.py

```diff
@@ -251,13 +252,31 @@ def permits(self, context, principals, permission):
         subpolicy_permits = self.policy.permits(context, principals, permission)
 
         # If the request is permitted by the subpolicy, check if the context is
-        # 2FA requireable, if 2FA is indeed required, and if the user has 2FA
+        # 2FA requireable, if 2FA is indeed required, and if the identity has 2FA
         # enabled
         if subpolicy_permits and isinstance(context, TwoFactorRequireable):
+            if isinstance(request.identity, User):
+                # If this is a user-identified request, then we can simply ask
+                # whether the user has 2FA enabled.
+                has_two_factor = request.identity.has_two_factor
+            elif isinstance(request.identity, OIDCPublisher):
+                # Otherwise, if it's an OIDC-identified request, then we can
+                # ask whether the context (a project) has users who all
+                # have 2FA enabled.
+                # NOTE: TwoFactorRequireable applies to models other than projects,
+                # but we can assume that the context is a project here because of
+                # our subpolicy: OIDC identities only appear with macaroons,
+                # and macaroons are only permitted on the `upload` permission
+                # with a project context.
+                has_two_factor = all(user.has_two_factor for user in context.users)
```

Comment:
Won't this result in publishing failing for the entire project if one user removes 2FA? I think that would be an unexpected outcome.

Given that a Publisher doesn't represent a specific user, I _think_ I'm OK skipping the check for 2FA in this case, assuming we are appropriately limiting OIDC Publisher creation for 2FA-mandated projects to users with 2FA only.